diff --git a/src/base/base_uses.f90 b/src/base/base_uses.f90
index 106072a187..4459734eb8 100644
--- a/src/base/base_uses.f90
+++ b/src/base/base_uses.f90
@@ -32,7 +32,12 @@
 ! In contrast to CPWARN, the warning counter is not increased
 #define CPHINT(msg) CALL cp__h(__SHORT_FILE__,__LINE__,msg)
 
+! CPASSERT can be elided if NDEBUG is defined.
+#if defined(NDEBUG)
+# define CPASSERT(cond)
+#else
 # define CPASSERT(cond) IF(.NOT.(cond))CALL cp__a(__SHORT_FILE__,__LINE__)
+#endif
 
 ! The MARK_USED macro can be used to mark an argument/variable as used. It is intended to make
 ! it possible to switch on -Werror=unused-dummy-argument, but deal elegantly with, e.g.,
